pkg/k8shandler/forwarding.go

```diff
@@ -75,197 +46,196 @@ func (clusterRequest *ClusterLoggingRequest) generateCollectorConfig() (config s
+			Name: pipeline.Name, InputRefs: goodIn.List(), OutputRefs: goodOut.List(),
+		})
+	}
+}
 
-		if len(outStatus.Conditions) == 0 {
-			outStatus.State = logforward.OutputStateAccepted
-			refs.Insert(output.Name)
-			outputs = append(outputs, output)
+func (clusterRequest *ClusterLoggingRequest) verifyOutputs(spec *logging.ClusterLogForwarderSpec, status *logging.ClusterLogForwarderStatus) {
+	status.Outputs = logging.NamedConditions{}
+	for i, output := range clusterRequest.ForwarderSpec.Outputs {
+		conds := status.Outputs.Get(output.Name)
+		badName := func(msg ...interface{}) {
+			output.Name = fmt.Sprintf("output[%v]", i)
+			conds = status.Outputs.Get(output.Name)
+			setInvalid(conds, msg...)
+		}
+
+		switch {
+		case output.Name == "":
+			badName("output must have a name")
+		case logging.ReservedOutputNames.Has(output.Name):
+			badName("output name %q is reserved", output.Name)
+		case len(conds) > 0:
+			badName("duplicate name: %q", output.Name)
+		case !logging.IsOutputTypeName(output.Type):
+			setInvalid(conds, "unknown output type: %q", output.Type)
+		case output.URL == "":
+			setInvalid(conds, "missing URL")
+		default:
+			clusterRequest.verifyOutputSecret(&output, conds)
+		}
+		if len(conds) == 0 {
+			conds.SetNew(logging.ConditionReady, true, "")
+			spec.Outputs = append(spec.Outputs, output)
+		}
+	}
+	// Add the default output if required and available.
+	routes := logging.NewRoutes(clusterRequest.ForwarderSpec.Pipelines)
+	if _, ok := routes.ByOutput[logging.OutputNameDefault]; ok {
+		conds := status.Outputs.Get(logging.OutputNameDefault)
+		if clusterRequest.cluster.Spec.LogStore == nil {
+			conds.SetNew(logging.ConditionReady, false, logging.ReasonMissingResource, "no default log store specified")
+		} else {
+			spec.Outputs = append(spec.Outputs, logging.OutputSpec{
+				Name:   logging.OutputNameDefault,
+				Type:   logging.OutputTypeElasticsearch,
+				URL:    constants.LogStoreURL,
+				Secret: &logging.OutputSecretSpec{Name: constants.CollectorSecretName},
+			})
+			conds.SetNew(logging.ConditionReady, true, "")
 		}
-		logger.Debugf("Status of output evaluation: %v", outStatus)
-		status.Outputs = append(status.Outputs, outStatus)
-
 	}
-	return refs, outputs
 }
 
-func verifyOutputSecret(output logforward.OutputSpec, secret *core.Secret, status *logforward.OutputStatus) {
-	if output.Type != logforward.OutputTypeForward || secret == nil {
+func (clusterRequest *ClusterLoggingRequest) verifyOutputSecret(output *logging.OutputSpec, conds logging.Conditions) {
+	if output.Secret == nil {
+		return
+	}
+	name := output.Secret.Name
+	if name == "" {
+		setInvalid(conds, "secretRef must have a name")
 		return
 	}
-	if _, exists := secret.Data["shared_key"]; !exists {
-		status.AddCondition(logforward.OutputConditionTypeSecret, logforward.OutputConditionReasonSecretMissingSharedKey, "")
+	if _, err := clusterRequest.GetSecret(name); err != nil {
```

Comment:
The err check should be limit to `errors.IsNotFound` or else we could just err on transient errors, e.g. networking errors. OTOH if we don't limit the err check we should reconcile the whole thing again. WDYT?

Reply:
the `GetSecret` function already does that check :)
https://github.com/openshift/cluster-logging-operator/blob/master/pkg/k8shandler/secret.go#L67-L77

Reply:
This does bring up another concern that we aren't doing the same comparison as before, where we ensure the `shared_key` exists within the secret...

Reply:
I'll restore the original logic - there's no good reason for the change.

Reply:
The secret can be used for other output types, not just secure fluent forwarding. Not all will have a shared_secret. Perhaps we add some validation routines per output-type to vet the secret. One for after the merge.